### PR TITLE
[Overview] Dependency Management RFC

### PR DESCRIPTION
**Pull request by sophiewigmore**
## Summary

[Readable](https://github.com/paketo-buildpacks/rfcs/blob/dependency-management-top-level/text/dependencies/rfcs/0000-dependency-management-overview.md)

Overview RFC of the three-phased approach to changing our dependency management system.
- Related to https://github.com/paketo-community/explorations/issues/8

## Checklist

* [X] I have viewed, signed, and submitted the Contributor License Agreement.
* [X] I have linked issue(s) that this PR should close using keywords or the Github UI (See [docs](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue))
* [ ] I have added an integration test, if necessary.
* [X] I have reviewed the [styleguide](https://github.com/paketo-buildpacks/community/blob/main/STYLEGUIDE.md) for guidance on my code quality.
* [X] I'm happy with the commit history on this PR (I have rebased/squashed as needed).

### CODE REVIEW

File: text/dependencies/rfcs/0000-dependency-management-overview.md
Comment (robdimsdale) [reply text/dependencies/rfcs/0000-dependency-management-overview.md:164]: Nice idea @joshuatcasey . I'm wondering how we could make that work in an offline environment? I think we'd have to instruct the buildpack operator to copy images from a public registry to their air-gapped one prior to using the buildpack. That seems quite convoluted.

Comment (joshuatcasey) [reply text/dependencies/rfcs/0000-dependency-management-overview.md:164]: @robdimsdale The offline configuration is tougher. In order to take this buildpack and its dependencies offline you would have to intentionally choose some dependency versions to bring offline and make those layers available via some sort of docker repo or pre-loading them into the docker daemon.

Comment (robdimsdale) [reply text/dependencies/rfcs/0000-dependency-management-overview.md:164]: Yes, exactly @joshuatcasey - going in this direction would require some pretty significant changes around distributing the image layers. This was one of the use-cases that [skopeo](https://github.com/containers/skopeo) was designed to solve, so it's certainly possible.

That being said, I'd prefer to keep things as simple as possible for now, so I would suggest not exploring this alternative path at this time.

File: text/dependencies/rfcs/0000-dependency-management-overview.md
Comment (ForestEckhardt) [inline text/dependencies/rfcs/0000-dependency-management-overview.md:124]:
```suggestion
1. Federated Model Adoption In Buildpacks: [Phase 1 RFC](https://github.com/paketo-buildpacks/rfcs/blob/dependency-management-step-one/text/dependencies/rfcs/0000-dependency-management-phase-one.md)
```

File: text/dependencies/rfcs/0000-dependency-management-overview.md
Comment (ekcasey) [inline text/dependencies/rfcs/0000-dependency-management-overview.md:116]: Is this really a compilation problem or is it a metadata problem?

For example our compilation of `node` runs fine on jammy apparently
```
docker run -it -v /tmp/paketo-node:/node ubuntu:jammy bash
root@eaa064b73682:/# cd node
root@eaa064b73682:/node# tar xzf node_v14.19.2_linux_x64_bionic_3f263e3e.tgz
root@eaa064b73682:/node# bin/node --help
Usage: node [options] [ script.js ] [arguments]
       node inspect [options] [ script.js | host:port ] [arguments]
```

And indeed dynamically links against the same (presumably stable) set of glilbc APIs as the upstream distribution
```
cnb@0e9092f382aa:/workspace$ ldd /layers/paketo-buildpacks_node-engine/node/bin/node
	linux-vdso.so.1 (0x00007ffd61cd1000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f4729087000)
	libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f4728cfe000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f4728960000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f4728748000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f4728529000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f4728138000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f472dc00000)
```

In cases where the results of compilation differ between stacks (e.g. nodejs for alpine or `nginx` probably) we will still need multiple version per stack even if we fetch binaries from from upstream.

Not suggesting that we shouldn't move away from recompiling, but that in some cases we could add support for more stacks with our current set of deps.

File: text/dependencies/rfcs/0000-dependency-management-overview.md
Comment (ekcasey) [inline text/dependencies/rfcs/0000-dependency-management-overview.md:102]: Can we approve one of these phases without the others? Seems like we can't approve the overview without approving the phases at least. In which case, I am not sure what gain from splitting the RFC up. I could definitely see us tracking implementation in three phases but maybe it would be better to have the entire plan (including the three phases) in a single document so we can approve it all in one go and reference the details in one place.

File: text/dependencies/rfcs/0000-dependency-management-overview.md
Comment (sophiewigmore) [reply text/dependencies/rfcs/0000-dependency-management-overview.md:146]: This is fair, I don't think the workflows will be too difficult to manage for Tooling maintainers, since most of the logic is going to be in the buildpack, and the workflows will just stick the steps together. I maintain Dependencies and Tooling though, so other maintainers should probably weigh in...

File: text/dependencies/rfcs/0000-dependency-management-overview.md
Comment (sophiewigmore) [reply text/dependencies/rfcs/0000-dependency-management-overview.md:102]: The phases are pretty dependent on each other, but I think there are some details that are independent. We could agree on Phase 1 (buildpack-related changes) and get that approved, and then go back to the Phase 2 RFC and hash out the details afterwards. The separation of RFCs is tied to the phased implementation, but I do think it makes the content more digestible by separating concerns. It's a lot of content to put into one RFC in my opinion.

Comment (sophiewigmore) [reply text/dependencies/rfcs/0000-dependency-management-overview.md:102]: It makes it easier for me as the RFC author to have a sort of separation of concerns, but there is some overlap that may make it confusing for others. My intention with the phases to make it clearer!

File: text/dependencies/rfcs/0000-dependency-management-overview.md
Comment (fg-j) [reply text/dependencies/rfcs/0000-dependency-management-overview.md:146]: I'm fine with owning the dependency workflows, as a tooling maintainer.

File: text/dependencies/rfcs/0000-dependency-management-overview.md
Comment (ryanmoran) [reply text/dependencies/rfcs/0000-dependency-management-overview.md:134]: Bump

File: text/dependencies/rfcs/0000-dependency-management-overview.md
Comment (sophiewigmore) [reply text/dependencies/rfcs/0000-dependency-management-overview.md:146]: Update the RFC to reflect this as we have 3/4 tooling maintainers with 👍

File: text/dependencies/rfcs/0000-dependency-management-overview.md
Comment (dmikusa-pivotal) [reply text/dependencies/rfcs/0000-dependency-management-overview.md:154]: At a high level, there's a desire to consolidate our pipelines. I don't know what the path to a consolidated pipeline would look like though. We'll probably have to go in steps and gradually get to it.

I'm +1 for sharing actions & code to check for updates. Our current code is a copy/pasta mess without any testing, so I'd be happy to retire it. 

I don't think that we could ditch pipeline-builder, at least not out of the gate, but we could probably have it generate pipelines that are inline with the update resource pipelines being proposed here. That would help to align on the actions and tooling we're using to update buildpack.toml. I do worry about the dependency metadata. We're not totally aligned on that, but I don't think it's so far off that we can't make something work either.

**Review by brayanhenao — APPROVED** [review]

File: text/dependencies/rfcs/0000-dependency-management-overview.md
Comment (fg-j) [reply text/dependencies/rfcs/0000-dependency-management-overview.md:134]: Bump!

Comment (sophiewigmore) [reply text/dependencies/rfcs/0000-dependency-management-overview.md]: done!

**Comment by joshuatcasey** [comment]
2022-07-26 Working Group notes:

@ryanmoran encourages merging, we have clarity around the broad strokes int he RFCs and we can go ahead and merge.

**Review by dmikusa-pivotal — APPROVED** [review]

**Review by joshuatcasey — APPROVED** [review]

**Review by ForestEckhardt — APPROVED** [review]

**Pull request merged**
